File: 백준/Silver/1874. 스택 수열/스택 수열.py
#  **************************************************************************  #
#                                                                              #
#                                                       :::    :::    :::      #
#    Problem Number: 1874                              :+:    :+:      :+:     #
#                                                     +:+    +:+        +:+    #
#    By: hrkim2001 <boj.kr/u/hrkim2001>              +#+    +#+          +#+   #
#                                                   +#+      +#+        +#+    #
#    https://boj.kr/1874                           #+#        #+#      #+#     #
#    Solved: 2025/08/18 02:44:47 by hrkim2001     ###          ###   ##.kr     #
#                                                                              #
#  **************************************************************************  #
n = int(input())
A = [int(input()) for i in range(n)]

stack = []
num = 1
result = True
answer = ''

for i in range(0, n):
    
    while A[i] >= num:
        stack.append(num)
        num += 1
        answer += '+\n'
    
    if A[i] == stack[-1]:
        stack.pop()
        answer += '-\n'
        continue
        
    if A[i] < stack[-1]:
        result = False
        print('NO')
        break

if result:
    print(answer)

File: 백준/Silver/1874. 스택 수열/스택 수열.java
/* ************************************************************************** */
/*                                                                            */
/*                                                      :::    :::    :::     */
/*   Problem Number: 1874                              :+:    :+:      :+:    */
/*                                                    +:+    +:+        +:+   */
/*   By: hrkim2001 <boj.kr/u/hrkim2001>              +#+    +#+          +#+  */
/*                                                  +#+      +#+        +#+   */
/*   https://boj.kr/1874                           #+#        #+#      #+#    */
/*   Solved: 2026/01/04 13:21:44 by hrkim2001     ###          ###   ##.kr    */
/*                                                                            */
/* ************************************************************************** */
import java.util.Scanner;
import java.util.Stack;

public class Main {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        int[] A = new int[n];
        for (int i = 0; i < n; i++) {
            A[i] = sc.nextInt();
        }

        Stack<Integer> stack = new Stack<>();
        StringBuffer bf = new StringBuffer();
        int num = 1;
        boolean result = true;

        for (int i = 0; i < n; i++) {
            while(stack.empty() || A[i] >= num) {
                stack.push(num);
                num++;
                bf.append("+\n");
            }

            int stackNum = stack.pop();
            if(stackNum < A[i]) {
                result = false;
                System.out.println("NO");
                break;
            } else{
                bf.append("-\n");
            }
        }
        if(result) {
            System.out.println(bf.toString());
        }
    }
}